providers/configcat/src/main/java/dev/openfeature/contrib/providers/configcat/ConfigCatProvider.java

```diff
@@ -0,0 +1,205 @@
+import lombok.Getter;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Provider implementation for ConfigCat.
+ */
+@Slf4j
+public class ConfigCatProvider extends EventProvider {
+
+    @Getter
+    private static final String NAME = "ConfigCat";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+
+    private ConfigCatProviderConfig configCatProviderConfig;
+
+    @Getter
+    private ConfigCatClient configCatClient;
+
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param configCatProviderConfig configCatProvider Config
+     */
+    public ConfigCatProvider(ConfigCatProviderConfig configCatProviderConfig) {
+        this.configCatProviderConfig = configCatProviderConfig;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        configCatClient = ConfigCatClient.get(configCatProviderConfig.getSdkKey(),
+            configCatProviderConfig.getOptions());
+        configCatProviderConfig.postInit();
+        state = ProviderState.READY;
+        log.info("finished initializing provider, state: {}", state);
+
+        configCatClient.getHooks().addOnClientReady(() -> {
+            state = ProviderState.READY;
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .message("provider ready")
+                .build();
+            emitProviderReady(providerEventDetails);
+        });
+
+        configCatClient.getHooks().addOnConfigChanged(map -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .flagsChanged(new ArrayList<>(map.keySet()))
+                .message("config changed")
+                .build();
+            emitProviderReady(providerEventDetails);
+        });
+
+        configCatClient.getHooks().addOnError(errorMessage -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .message(errorMessage)
+                .build();
+            emitProviderError(providerEventDetails);
+        });
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
```

Comment:
As all the evaluations are basically doing the same, I'd introduce a generic evaluate function and call it from the getBooleanEvaluation, getStringEvaluation, getIntegerEvaluation, getDoubleEvaluation.

Reply:
generally it's right, here the evaluate is 1-2 actions, with changed class. When trying to call _configCatClient
            .getValueDetails(Class_ with generic type, it does [not work in a simple way](https://stackoverflow.com/questions/3437897/how-do-i-get-a-class-instance-of-generic-type-t), to make it work make it less clean.

Reply:
I'm fine with the current solution too. 
I was thinking on something like this, it seems much cleaner to me:
 
```
    @Override
    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
        return evaluate(Boolean.class, key, defaultValue, ctx);
    }

    @Override
    public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue, EvaluationContext ctx) {
        return evaluate(String.class, key, defaultValue, ctx);
    }

    @Override
    public ProviderEvaluation<Double> getDoubleEvaluation(String key, Double defaultValue, EvaluationContext ctx) {
        return evaluate(Double.class, key, defaultValue, ctx);
    }

    @Override
    public ProviderEvaluation<Integer> getIntegerEvaluation(String key, Integer defaultValue, EvaluationContext ctx) {
        return evaluate(Integer.class, key, defaultValue, ctx);
    }

    private <T> ProviderEvaluation<T> evaluate(Class<T> classOfT, String key, T defaultValue, EvaluationContext ctx) {
        User user = ctx == null ? null : dev.openfeature.contrib.providers.configcat.ContextTransformer.transform(ctx);
        EvaluationDetails<T> evaluationDetails = configCatClient
            .getValueDetails(classOfT, key, user, defaultValue);
        return ProviderEvaluation.<T>builder()
            .value(evaluationDetails.getValue())
            .variant(evaluationDetails.getVariationId())
            .errorMessage(evaluationDetails.getError())
            .build();
    }
```

Reply:
Done, except _getObjectEvaluation_ which evaluates a String, and returns _Value_.

Reply:
This is a nice optimization.